File: extensions/db-bridge/drivers/postgres.js
const { Client } = require('pg');

module.exports = {
  // Store the active client instance
  client: null,

  /**
   * Connect to the PostgreSQL database
   * @param {Object} config Connection config (host, port, user, password, database)
   */
  connect: async (config) => {
    // If there's an existing client, disconnect it first
    if (module.exports.client) {
      await module.exports.disconnect();
    }

    const client = new Client({
      host: config.host || 'localhost',
      port: config.port ? parseInt(config.port, 10) : 5432,
      user: config.user || config.username,
      password: config.password,
      database: config.database,
      ssl: config.ssl ? { rejectUnauthorized: false } : false,
      connectionTimeoutMillis: 5000 // 5 seconds timeout
    });

    await client.connect();
    module.exports.client = client;
    return true;
  },

  /**
   * Execute an arbitrary SQL query
   * @param {string} sql The SQL query string
   */
  query: async (sql) => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }
    
    const result = await module.exports.client.query(sql);
    return {
      rows: result.rows,
      fields: result.fields.map(f => ({ name: f.name, dataTypeID: f.dataTypeID })),
      rowCount: result.rowCount
    };
  },

  /**
   * Get the database schema (tables, views, functions)
   */
  getSchema: async () => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }

    // Tables
    const tablesResult = await module.exports.client.query(`
      SELECT table_name 
      FROM information_schema.tables 
      WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
      ORDER BY table_name;
    `);

    // Views
    const viewsResult = await module.exports.client.query(`
      SELECT table_name 
      FROM information_schema.views 
      WHERE table_schema = 'public'
      ORDER BY table_name;
    `);

    // Functions
    const functionsResult = await module.exports.client.query(`
      SELECT routine_name 
      FROM information_schema.routines 
      WHERE routine_schema = 'public' AND routine_type = 'FUNCTION'
      ORDER BY routine_name;
    `);

    return {
      tables: tablesResult.rows.map(r => ({ name: r.table_name })),
      views: viewsResult.rows.map(r => ({ name: r.table_name })),
      functions: functionsResult.rows.map(r => ({ name: r.routine_name })),
      schemas: [{ name: 'public' }]
    };
  },

  /**
   * Fetch data from a specific table with pagination
   */
  fetchTableData: async (tableName, limit = 100, offset = 0, sortColumn = null, sortDirection = 'asc', filter = null) => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }
    
    const safeTable = `"${tableName.replace(/"/g, '""')}"`;
    
    // Build WHERE clause if filter is specified
    let whereClause = '';
    const queryParams = [limit, offset];
    if (filter) {
      whereClause = ` WHERE CAST(t.* AS TEXT) ILIKE $3`;
      queryParams.push(`%${filter}%`);
    }
    
    // Find primary keys for the table
    const pkQuery = `
      SELECT kcu.column_name
      FROM information_schema.table_constraints tco
      JOIN information_schema.key_column_usage kcu 
           ON kcu.constraint_name = tco.constraint_name
           AND kcu.constraint_schema = tco.constraint_schema
      WHERE tco.constraint_type = 'PRIMARY KEY' 
        AND kcu.table_name = $1
        AND tco.table_schema = 'public';
    `;
    
    // Build ORDER BY clause if sort is specified
    let orderClause = '';
    if (sortColumn) {
      const safeCol = `"${sortColumn.replace(/"/g, '""')}"`;
      const dir = sortDirection === 'desc' ? 'DESC' : 'ASC';
      orderClause = ` ORDER BY ${safeCol} ${dir}`;
    }
    
    const query = `SELECT * FROM public.${safeTable} t${whereClause}${orderClause} LIMIT $1 OFFSET $2;`;
    const countQuery = `SELECT COUNT(*) as total FROM public.${safeTable} t${whereClause};`;
    
    const pkResult = await module.exports.client.query(pkQuery, [tableName]);
    const dataResult = await module.exports.client.query(query, queryParams);
    
    // For count query, if there's a filter, it will be the third param in whereClause ($3)
    // but the count query doesn't have $1 and $2 (limit/offset).
    // So we replace $3 with $1 in the count query.
    const finalCountQuery = filter ? countQuery.replace('$3', '$1') : countQuery;
    const finalCountParams = filter ? [queryParams[2]] : [];
    const countResult = await module.exports.client.query(finalCountQuery, finalCountParams);
    
    const pkColumns = pkResult.rows.map(r => r.column_name);

    return {
      rows: dataResult.rows,
      fields: dataResult.fields.map(f => ({ 
        name: f.name, 
        dataTypeID: f.dataTypeID,
        isPrimaryKey: pkColumns.includes(f.name)
      })),
      totalCount: parseInt(countResult.rows[0].total, 10)
    };
  },

  /**
   * Update a specific cell in a table
   */
  updateCell: async (tableName, pkColumn, pkValue, targetColumn, newValue) => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }
    const safeTable = `"${tableName.replace(/"/g, '""')}"`;
    const safePkCol = `"${pkColumn.replace(/"/g, '""')}"`;
    const safeTargetCol = `"${targetColumn.replace(/"/g, '""')}"`;
    
    const query = `UPDATE public.${safeTable} SET ${safeTargetCol} = $1 WHERE ${safePkCol} = $2`;
    await module.exports.client.query(query, [newValue, pkValue]);
    return true;
  },

  /**
   * Insert a new row into a table
   * @param {string} tableName
   * @param {Object} data  { colName: value, … }
   * Returns the inserted row (with DB-generated defaults like serial IDs).
   */
  insertRow: async (tableName, data) => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }
    const safeTable = `"${tableName.replace(/"/g, '""')}"`;
    const cols = Object.keys(data);
    if (cols.length === 0) {
      // INSERT with all defaults
      const result = await module.exports.client.query(
        `INSERT INTO public.${safeTable} DEFAULT VALUES RETURNING *`
      );
      return result.rows[0];
    }
    const safeCols = cols.map(c => `"${c.replace(/"/g, '""')}"`).join(', ');
    const placeholders = cols.map((_, i) => `$${i + 1}`).join(', ');
    const values = cols.map(c => data[c]);
    
    const query = `INSERT INTO public.${safeTable} (${safeCols}) VALUES (${placeholders}) RETURNING *`;
    console.log('[Postgres Driver] insertRow Executing:', query, 'with values:', values);
    
    try {
      const result = await module.exports.client.query(query, values);
      return result.rows[0];
    } catch (err) {
      throw new Error(`[Query: ${query} | Values: ${JSON.stringify(values)}] ${err.message}`);
    }
  },

  /**
   * Delete rows from a table by primary key values
   * @param {string} tableName
   * @param {string} pkColumn
   * @param {Array}  pkValues  Array of PK values to delete
   */
  deleteRows: async (tableName, pkColumn, pkValues) => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }
    const safeTable = `"${tableName.replace(/"/g, '""')}"`;
    const safePk = `"${pkColumn.replace(/"/g, '""')}"`;
    const placeholders = pkValues.map((_, i) => `$${i + 1}`).join(', ');
    await module.exports.client.query(
      `DELETE FROM public.${safeTable} WHERE ${safePk} = ANY(ARRAY[${placeholders}])`,
      pkValues
    );
    return true;
  },

  /**
   * Export table data to CSV file
   */
  exportToCSV: async (tableName, exportPath) => {
    if (!module.exports.client) {
      throw new Error("Not connected to database");
    }
    const fs = require('fs');
    const safeTable = `"${tableName.replace(/"/g, '""')}"`;
    
    const writeStream = fs.createWriteStream(exportPath);
    
    // Write header
    const schemaResult = await module.exports.client.query(`SELECT * FROM public.${safeTable} LIMIT 0`);
    const headers = schemaResult.fields.map(f => f.name);
    writeStream.write(headers.join(',') + '\n');
    
    // Simple batching to avoid memory issues on large tables
    let offset = 0;
    const batchSize = 2000;
    let hasMore = true;
    
    while (hasMore) {
      const result = await module.exports.client.query(`SELECT * FROM public.${safeTable} LIMIT $1 OFFSET $2`, [batchSize, offset]);
      if (result.rows.length === 0) {
        hasMore = false;
        break;
      }
      
      for (const row of result.rows) {
        const line = headers.map(h => {
          let val = row[h];
          if (val === null || val === undefined) return '';
          val = String(val).replace(/"/g, '""');
          if (val.includes(',') || val.includes('\\n') || val.includes('"')) {
            return `"${val}"`;
          }
          return val;
        }).join(',');
        writeStream.write(line + '\n');
      }
      offset += batchSize;
    }
    
    return new Promise((resolve, reject) => {
      writeStream.end(() => resolve(true));
      writeStream.on('error', reject);
    });
  },

  /**
   * Get detailed column definitions for a table
   */
  getTableColumns: async (tableName) => {
    if (!module.exports.client) throw new Error("Not connected to database");
    
    // Fetch columns with metadata
    const query = `
      SELECT 
        c.column_name as name,
        c.data_type as "dataType",
        c.is_nullable as nullable,
        c.column_default as "default"
      FROM information_schema.columns c
      WHERE c.table_schema = 'public' AND c.table_name = $1
      ORDER BY c.ordinal_position;
    `;
    const result = await module.exports.client.query(query, [tableName]);
    return result.rows.map(r => ({
      name: r.name,
      dataType: r.dataType,
      nullable: r.nullable === 'YES',
      default: r.default
    }));
  },

  /**
   * Alter table (ADD, DROP, RENAME columns)
   */
  alterTable: async (tableName, operations) => {
    if (!module.exports.client) throw new Error("Not connected to database");
    const safeTable = `"${tableName.replace(/"/g, '""')}"`;

    // Process each operation sequentially
    for (const op of operations) {
      if (op.type === 'ADD_COLUMN') {
        const safeCol = `"${op.name.replace(/"/g, '""')}"`;
        let def = `ADD COLUMN ${safeCol} ${op.dataType}`;
        if (op.nullable === false) def += ' NOT NULL';
        if (op.default !== undefined && op.default !== null && op.default !== '') {
          // simple check for numbers/functions vs strings
          const isNumericOrFunc = /^[0-9]+$/.test(op.default) || op.default.includes('(');
          def += ` DEFAULT ${isNumericOrFunc ? op.default : "'" + op.default.replace(/'/g, "''") + "'"}`;
        }
        await module.exports.client.query(`ALTER TABLE public.${safeTable} ${def}`);
      } 
      else if (op.type === 'DROP_COLUMN') {
        const safeCol = `"${op.name.replace(/"/g, '""')}"`;
        await module.exports.client.query(`ALTER TABLE public.${safeTable} DROP COLUMN ${safeCol}`);
      } 
      else if (op.type === 'RENAME_COLUMN') {
        const safeOld = `"${op.oldName.replace(/"/g, '""')}"`;
        const safeNew = `"${op.newName.replace(/"/g, '""')}"`;
        await module.exports.client.query(`ALTER TABLE public.${safeTable} RENAME COLUMN ${safeOld} TO ${safeNew}`);
      }
    }
    return true;
  },

  /**
   * Disconnect from the database
   */
  disconnect: async () => {
    if (module.exports.client) {
      try {
        await module.exports.client.end();
      } catch (err) {
        console.error("Error disconnecting from postgres:", err);
      } finally {
        module.exports.client = null;
      }
    }
  }
};

File: extensions/db-bridge/internal/drivers/postgres.go
package drivers

import (
	"database/sql"
	"fmt"
	"strings"

	_ "github.com/jackc/pgx/v5/stdlib"
)

type PostgresDriver struct {
	db *sql.DB
}

func (d *PostgresDriver) Connect(config Config) error {
	sslMode := "disable"
	if config.SSL {
		sslMode = "require"
	}

	dsn := fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=%s&default_query_exec_mode=simple_protocol",
		config.Username, config.Password, config.Host, config.Port, config.Database, sslMode)

	db, err := sql.Open("pgx", dsn)
	if err != nil {
		return err
	}

	if err := db.Ping(); err != nil {
		return err
	}

	d.db = db
	return nil
}

func (d *PostgresDriver) Disconnect() error {
	if d.db != nil {
		return d.db.Close()
	}
	return nil
}

func (d *PostgresDriver) GetSchema() (*SchemaResult, error) {
	tablesRows, err := d.db.Query(`
		SELECT table_name 
		FROM information_schema.tables 
		WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
		ORDER BY table_name;
	`)
	if err != nil {
		return nil, err
	}
	defer tablesRows.Close()

	var tables []SchemaObject = []SchemaObject{}
	for tablesRows.Next() {
		var name string
		if err := tablesRows.Scan(&name); err != nil {
			return nil, err
		}
		tables = append(tables, SchemaObject{Name: name})
	}

	viewsRows, err := d.db.Query(`
		SELECT table_name 
		FROM information_schema.views 
		WHERE table_schema = 'public'
		ORDER BY table_name;
	`)
	if err != nil {
		return nil, err
	}
	defer viewsRows.Close()

	var views []SchemaObject = []SchemaObject{}
	for viewsRows.Next() {
		var name string
		if err := viewsRows.Scan(&name); err != nil {
			return nil, err
		}
		views = append(views, SchemaObject{Name: name})
	}

	funcsRows, err := d.db.Query(`
		SELECT routine_name 
		FROM information_schema.routines 
		WHERE routine_schema = 'public' AND routine_type = 'FUNCTION'
		ORDER BY routine_name;
	`)
	if err != nil {
		return nil, err
	}
	defer funcsRows.Close()

	var funcs []SchemaObject = []SchemaObject{}
	for funcsRows.Next() {
		var name string
		if err := funcsRows.Scan(&name); err != nil {
			return nil, err
		}
		funcs = append(funcs, SchemaObject{Name: name})
	}

	return &SchemaResult{
		Tables:    tables,
		Views:     views,
		Functions: funcs,
		Schemas:   []SchemaObject{{Name: "public"}},
	}, nil
}

func (d *PostgresDriver) FetchTableData(tableName string, limit, offset int, sortColumn, sortDirection string) (*TableDataResult, error) {
	safeTable := fmt.Sprintf("\"%s\"", strings.ReplaceAll(tableName, "\"", "\"\""))
	
	// Get PKs
	pkRows, err := d.db.Query(`
		SELECT kcu.column_name
		FROM information_schema.table_constraints tco
		JOIN information_schema.key_column_usage kcu 
			ON kcu.constraint_name = tco.constraint_name
			AND kcu.constraint_schema = tco.constraint_schema
		WHERE tco.constraint_type = 'PRIMARY KEY' 
			AND kcu.table_name = $1
			AND tco.table_schema = 'public';
	`, tableName)
	if err != nil {
		return nil, err
	}
	defer pkRows.Close()

	pkMap := make(map[string]bool)
	for pkRows.Next() {
		var col string
		if err := pkRows.Scan(&col); err != nil {
			return nil, err
		}
		pkMap[col] = true
	}

	orderClause := ""
	if sortColumn != "" {
		dir := "ASC"
		if strings.ToUpper(sortDirection) == "DESC" {
			dir = "DESC"
		}
		orderClause = fmt.Sprintf(" ORDER BY \"%s\" %s", strings.ReplaceAll(sortColumn, "\"", "\"\""), dir)
	}

	query := fmt.Sprintf("SELECT * FROM public.%s%s LIMIT $1 OFFSET $2", safeTable, orderClause)
	rows, err := d.db.Query(query, limit, offset)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	data, fields, err := ScanRows(rows)
	if err != nil {
		return nil, err
	}

	for i := range fields {
		if pkMap[fields[i].Name] {
			fields[i].IsPrimaryKey = true
		}
	}

	countQuery := fmt.Sprintf("SELECT COUNT(*) FROM public.%s", safeTable)
	var total int
	if err := d.db.QueryRow(countQuery).Scan(&total); err != nil {
		return nil, err
	}

	return &TableDataResult{
		Rows:       data,
		Fields:     fields,
		TotalCount: total,
	}, nil
}

func (d *PostgresDriver) Query(sql string) (*QueryResult, error) {
	rows, err := d.db.Query(sql)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	data, fields, err := ScanRows(rows)
	if err != nil {
		return nil, err
	}

	return &QueryResult{
		Rows:     data,
		Fields:   fields,
		RowCount: len(data),
	}, nil
}

func (d *PostgresDriver) UpdateCell(tableName, pkColumn string, pkValue interface{}, targetColumn string, newValue interface{}) error {
	safeTable := fmt.Sprintf("\"%s\"", strings.ReplaceAll(tableName, "\"", "\"\""))
	safePk := fmt.Sprintf("\"%s\"", strings.ReplaceAll(pkColumn, "\"", "\"\""))
	safeTarget := fmt.Sprintf("\"%s\"", strings.ReplaceAll(targetColumn, "\"", "\"\""))

	query := fmt.Sprintf("UPDATE public.%s SET %s = $1 WHERE %s = $2", safeTable, safeTarget, safePk)
	_, err := d.db.Exec(query, newValue, pkValue)
	return err
}

func (d *PostgresDriver) InsertRow(tableName string, data map[string]interface{}) (map[string]interface{}, error) {
	safeTable := fmt.Sprintf("\"%s\"", strings.ReplaceAll(tableName, "\"", "\"\""))
	
	if len(data) == 0 {
		query := fmt.Sprintf("INSERT INTO public.%s DEFAULT VALUES RETURNING *", safeTable)
		rows, err := d.db.Query(query)
		if err != nil {
			return nil, err
		}
		defer rows.Close()
		res, _, err := ScanRows(rows)
		if err != nil || len(res) == 0 {
			return nil, err
		}
		return res[0], nil
	}

	cols := make([]string, 0, len(data))
	values := make([]interface{}, 0, len(data))
	placeholders := make([]string, 0, len(data))
	
	i := 1
	for k, v := range data {
		cols = append(cols, fmt.Sprintf("\"%s\"", strings.ReplaceAll(k, "\"", "\"\"")))
		values = append(values, v)
		placeholders = append(placeholders, fmt.Sprintf("$%d", i))
		i++
	}

	query := fmt.Sprintf("INSERT INTO public.%s (%s) VALUES (%s) RETURNING *", 
		safeTable, strings.Join(cols, ", "), strings.Join(placeholders, ", "))
	
	rows, err := d.db.Query(query, values...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	res, _, err := ScanRows(rows)
	if err != nil || len(res) == 0 {
		return nil, err
	}
	return res[0], nil
}

func (d *PostgresDriver) DeleteRows(tableName, pkColumn string, pkValues []interface{}) error {
	safeTable := fmt.Sprintf("\"%s\"", strings.ReplaceAll(tableName, "\"", "\"\""))
	safePk := fmt.Sprintf("\"%s\"", strings.ReplaceAll(pkColumn, "\"", "\"\""))
	
	placeholders := make([]string, len(pkValues))
	for i := range pkValues {
		placeholders[i] = fmt.Sprintf("$%d", i+1)
	}

	query := fmt.Sprintf("DELETE FROM public.%s WHERE %s IN (%s)", 
		safeTable, safePk, strings.Join(placeholders, ", "))
	
	_, err := d.db.Exec(query, pkValues...)
	return err
}

func (d *PostgresDriver) GetTableColumns(tableName string) ([]TableColumn, error) {
	rows, err := d.db.Query(`
		SELECT 
			column_name, data_type, is_nullable, column_default
		FROM information_schema.columns 
		WHERE table_schema = 'public' AND table_name = $1
		ORDER BY ordinal_position;
	`, tableName)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var cols []TableColumn
	for rows.Next() {
		var c TableColumn
		var nullable string
		if err := rows.Scan(&c.Name, &c.DataType, &nullable, &c.Default); err != nil {
			return nil, err
		}
		c.Nullable = nullable == "YES"
		cols = append(cols, c)
	}
	return cols, nil
}

func (d *PostgresDriver) AlterTable(tableName string, operations []AlterOperation) error {
	safeTable := fmt.Sprintf("\"%s\"", strings.ReplaceAll(tableName, "\"", "\"\""))
	
	for _, op := range operations {
		var query string
		switch op.Type {
		case "ADD_COLUMN":
			safeCol := fmt.Sprintf("\"%s\"", strings.ReplaceAll(op.Name, "\"", "\"\""))
			query = fmt.Sprintf("ALTER TABLE public.%s ADD COLUMN %s %s", safeTable, safeCol, op.DataType)
			if op.Nullable != nil && !*op.Nullable {
				query += " NOT NULL"
			}
			if op.Default != nil {
				query += fmt.Sprintf(" DEFAULT %v", op.Default)
			}
		case "DROP_COLUMN":
			safeCol := fmt.Sprintf("\"%s\"", strings.ReplaceAll(op.Name, "\"", "\"\""))
			query = fmt.Sprintf("ALTER TABLE public.%s DROP COLUMN %s", safeTable, safeCol)
		case "RENAME_COLUMN":
			safeOld := fmt.Sprintf("\"%s\"", strings.ReplaceAll(op.OldName, "\"", "\"\""))
			safeNew := fmt.Sprintf("\"%s\"", strings.ReplaceAll(op.NewName, "\"", "\"\""))
			query = fmt.Sprintf("ALTER TABLE public.%s RENAME COLUMN %s TO %s", safeTable, safeOld, safeNew)
		}
		
		if query != "" {
			if _, err := d.db.Exec(query); err != nil {
				return err
			}
		}
	}
	return nil
}

func (d *PostgresDriver) ExportToCSV(tableName, exportPath string) error {
	return ExportToCSVHelper(d.db, "public.\""+strings.ReplaceAll(tableName, "\"", "\"\"")+"\"", exportPath, "\"")
}
